### PR TITLE
Use ChunkedSliceOutput to chunk and buffer writes in parquet

### PR DESCRIPTION
**Pull request by raunaqmorarka**
## Description
Avoids separate calls to output stream for writing page headers and page data.
Avoids holding on to extra memory for each buffered page due to over-allocation by compressors.
Adds memory usage accounting for buffered pages.

## Additional context and related issues
Should help with the memory usage accounting problems described in https://github.com/trinodb/trino/issues/18557

## Release notes

( ) This is not user-visible or is docs only, and no release notes are required.
( ) Release notes are required. Please propose a release note for me.
(x) Release notes are required, with the following suggested text:

```markdown
# Hive, Hudi, Iceberg, Delta
* Improve memory usage accounting of parquet writer. ({issue}`18564`)
```

### CODE REVIEW

File: lib/trino-parquet/src/main/java/io/trino/parquet/writer/PrimitiveColumnWriter.java [hunk 1]
```diff
@@ -257,21 +256,21 @@ private List<ParquetDataOutput> getDataStreams()
...
             byte[] pageBytes = dictionaryPage.getBytes().toByteArray();
             ParquetDataOutput pageData = compressor != null
                     ? compressor.compress(pageBytes)
                     : createDataOutput(Slices.wrappedBuffer(pageBytes));
-            long compressedSize = pageData.size();
+            int compressedSize = pageData.size();
 
             ByteArrayOutputStream dictStream = new ByteArrayOutputStream();
             parquetMetadataConverter.writeDictionaryPageHeader(
-                    toIntExact(uncompressedSize),
```
Comment (findepi) [inline lib/trino-parquet/src/main/java/io/trino/parquet/writer/PrimitiveColumnWriter.java:269]: > Allows simplifying some of the code working with pages sizes
to use ints instead of longs

I am not sure this is the right direction. Why `int` is better for data size?

At least i think https://github.com/trinodb/trino/pull/18479 is going in the opposite direction. (now added you as a reviewer there)

**Comment by findepi** [comment]
> Avoids separate calls to output stream for writing page headers and page data.

not sure i understand the goal here.
are we saving on http requests? i think the output stream is buffering internally?

can you expand a bit the commit rationale so that it's more obvious what's the objective?

File: lib/trino-parquet/src/main/java/io/trino/parquet/writer/PrimitiveColumnWriter.java (hunk 1 above)
Comment (raunaqmorarka) [reply lib/trino-parquet/src/main/java/io/trino/parquet/writer/PrimitiveColumnWriter.java:269]: All the underlying values here are ints, it's the usage of `long BytesInput#size` from parquet lib that was forcing us to use longs and then cast them back to int earlier. Now that has been cleaned up and it's no longer necessary to use a long. Everything implementing `ParquetDataOutput#size` is now actually returning an `int`, so storing it in a long and then casting to int is unnecessary.

**Comment by raunaqmorarka** [comment]
> > Avoids separate calls to output stream for writing page headers and page data.
> 
> not sure i understand the goal here. are we saving on http requests? i think the output stream is buffering internally?
> 
> can you expand a bit the commit rationale so that it's more obvious what's the objective?

Yes, there is varying levels of buffering in different output stream implementations. It's not clear to me that every implementation is doing that or that we should rely on that always being the case. In any case, that's not the main benefit of this change.

The main objective here is
> Avoids holding on to extra memory for each buffered page due to over-allocation by compressors.
> Adds memory usage accounting for buffered pages.

Also, ORC writer uses `ChunkedSliceOutput` in `OrcOutputBuffer#writeChunkToOutputStream` in a similar way.

File: lib/trino-parquet/src/main/java/io/trino/parquet/writer/PrimitiveColumnWriter.java (hunk 1 above)
Comment (findepi) [reply lib/trino-parquet/src/main/java/io/trino/parquet/writer/PrimitiveColumnWriter.java:269]: are we limited to 2 GB?

Comment (raunaqmorarka) [reply lib/trino-parquet/src/main/java/io/trino/parquet/writer/PrimitiveColumnWriter.java:269]: yup, we have a check in ParquetWriterConfig to limit `parquet.writer.block-size` to 2GB. The standard itself allows larger row groups, but it is not something we want users to do.

File: lib/trino-parquet/src/main/java/io/trino/parquet/writer/PrimitiveColumnWriter.java [hunk 2]
```diff
@@ -57,6 +57,8 @@ public class PrimitiveColumnWriter
         implements ColumnWriter
 {
     private static final int INSTANCE_SIZE = instanceSize(PrimitiveColumnWriter.class);
+    private static final int MINIMUM_OUTPUT_BUFFER_CHUNK_SIZE = 4 * 1024;
+    private static final int MAXIMUM_OUTPUT_BUFFER_CHUNK_SIZE = 1024 * 1024;
```
Comment (lukasz-stec) [inline lib/trino-parquet/src/main/java/io/trino/parquet/writer/PrimitiveColumnWriter.java:61]: any specific reasons for those constants? maybe comment?

File: lib/trino-parquet/src/main/java/io/trino/parquet/writer/ParquetWriter.java [hunk 3]
```diff
@@ -203,6 +203,7 @@ public void close()
         try (outputStream) {
             columnWriters.forEach(ColumnWriter::close);
             flush();
+            columnWriters = ImmutableList.of();
```
Comment (lukasz-stec) [inline lib/trino-parquet/src/main/java/io/trino/parquet/writer/ParquetWriter.java:206]: is this related to the memory accounting?

File: lib/trino-parquet/src/main/java/io/trino/parquet/writer/PrimitiveColumnWriter.java [hunk 4]
```diff
@@ -297,6 +297,7 @@ public long getBufferedBytes()
     public long getRetainedBytes()
     {
         return INSTANCE_SIZE +
+                compressedOutputStream.getRetainedSize() +
```
Comment (lukasz-stec) [inline lib/trino-parquet/src/main/java/io/trino/parquet/writer/PrimitiveColumnWriter.java:300]: does testing this make sense?

**Review by lukasz-stec — APPROVED** [review]
lgtm % comments

File: lib/trino-parquet/src/main/java/io/trino/parquet/writer/ParquetWriter.java (hunk 3 above)
Comment (raunaqmorarka) [reply lib/trino-parquet/src/main/java/io/trino/parquet/writer/ParquetWriter.java:206]: Yes, now that we are accounting for buffered pages I wanted to ensure that we release them and reduce retained memory as soon as possible. Buffered pages have to be held until they are flushed to output stream, so this is the most straightforward way to release them.
I think this memory would also get released by the current code in connector page sink calling commit on writer and then releasing reference to it. But that is a few layers above and would delay the cleanup a bit.

File: lib/trino-parquet/src/main/java/io/trino/parquet/writer/PrimitiveColumnWriter.java (hunk 2 above)
Comment (raunaqmorarka) [reply lib/trino-parquet/src/main/java/io/trino/parquet/writer/PrimitiveColumnWriter.java:61]: This is copied from ORC writer
I think the exact numbers don't matter that much here, just want to avoid too big byte arrays and too small chunks for writing to output stream

File: lib/trino-parquet/src/main/java/io/trino/parquet/writer/ParquetWriter.java (hunk 3 above)
Comment (lukasz-stec) [reply lib/trino-parquet/src/main/java/io/trino/parquet/writer/ParquetWriter.java:206]: ok, any reason to use an empty list instead of `null` (it starts as `null`)?

File: lib/trino-parquet/src/main/java/io/trino/parquet/writer/PrimitiveColumnWriter.java (hunk 4 above)
Comment (raunaqmorarka) [reply lib/trino-parquet/src/main/java/io/trino/parquet/writer/PrimitiveColumnWriter.java:300]: added a test

File: lib/trino-parquet/src/main/java/io/trino/parquet/writer/ParquetWriter.java (hunk 3 above)
Comment (raunaqmorarka) [reply lib/trino-parquet/src/main/java/io/trino/parquet/writer/ParquetWriter.java:206]: constructor calls `initColumnWriters` which ensures that it is never null, there are many usages of this field without a null check

File: lib/trino-parquet/src/main/java/io/trino/parquet/writer/ParquetDataOutput.java [hunk 5]
```diff
@@ -42,6 +43,25 @@ public void writeData(SliceOutput sliceOutput)
...
+            @Override
+            public int size()
+            {
+                return chunkedSliceOutput.size();
+            }
+
+            @Override
+            public void writeData(SliceOutput sliceOutput)
+            {
+                chunkedSliceOutput.getSlices().forEach(sliceOutput::writeBytes);
```
Comment (sopel39) [inline lib/trino-parquet/src/main/java/io/trino/parquet/writer/ParquetDataOutput.java:60]: > Avoids separate calls to output stream for writing page headers
and page data

Does it really avoid separate output calls if each chink is written separately?

File: lib/trino-parquet/src/main/java/io/trino/parquet/writer/PrimitiveColumnWriter.java [hunk 6]
```diff
@@ -249,7 +251,7 @@ private void flushCurrentPageToBuffer()
     private List<ParquetDataOutput> getDataStreams()
             throws IOException
     {
-        List<ParquetDataOutput> dictPage = new ArrayList<>();
+        ImmutableList.Builder<ParquetDataOutput> outputStreams = ImmutableList.builder();
```
Comment (sopel39) [inline lib/trino-parquet/src/main/java/io/trino/parquet/writer/PrimitiveColumnWriter.java:254]: why rename?

just `outputs`?

File: lib/trino-parquet/src/main/java/io/trino/parquet/writer/PrimitiveColumnWriter.java [hunk 7]
```diff
@@ -297,6 +297,7 @@ public long getBufferedBytes()
     public long getRetainedBytes()
```
Comment (sopel39) [inline lib/trino-parquet/src/main/java/io/trino/parquet/writer/PrimitiveColumnWriter.java:297]: Is it called after a `flush`? (is there something like a flush here)?

**Review by sopel39 — APPROVED** [review]

File: lib/trino-parquet/src/main/java/io/trino/parquet/writer/ParquetDataOutput.java [hunk 8]
```diff
@@ -42,6 +43,25 @@ public void writeData(SliceOutput sliceOutput)
...
+        return new ParquetDataOutput()
+        {
+            @Override
+            public int size()
+            {
+                return chunkedSliceOutput.size();
+            }
+
+            @Override
+            public void writeData(SliceOutput sliceOutput)
```
Comment (sopel39) [inline lib/trino-parquet/src/main/java/io/trino/parquet/writer/ParquetDataOutput.java:58]: I wonder do we actually need a copy now that we use `ChunkedSliceOutput `? 

Could we retain buffer and adjust it adaptively to avoid extra copy?

File: lib/trino-parquet/src/main/java/io/trino/parquet/writer/ParquetDataOutput.java (hunk 5 above)
Comment (raunaqmorarka) [reply lib/trino-parquet/src/main/java/io/trino/parquet/writer/ParquetDataOutput.java:60]: Earlier it was separate slices for page header and data, now that the min chunk size is above page header size (few bytes), we get bigger chunks for output stream.

File: lib/trino-parquet/src/main/java/io/trino/parquet/writer/PrimitiveColumnWriter.java (hunk 7 above)
Comment (raunaqmorarka) [reply lib/trino-parquet/src/main/java/io/trino/parquet/writer/PrimitiveColumnWriter.java:297]: There is no "flush" here, there is a "close" followed by a `getBuffer` to extract the buffered pages.
This is called by connector page sink for every writer after writing each page.

File: lib/trino-parquet/src/main/java/io/trino/parquet/writer/ParquetDataOutput.java (hunk 8 above)
Comment (sopel39) [reply lib/trino-parquet/src/main/java/io/trino/parquet/writer/ParquetDataOutput.java:58]: Anyway, this could be future improvment

Comment (raunaqmorarka) [reply lib/trino-parquet/src/main/java/io/trino/parquet/writer/ParquetDataOutput.java:58]: If you mean reusing the compression buffer byte array between pages, then yes I am experimenting with that separately. That only saves on new array allocations and doesn't avoid copying.

Comment (sopel39) [reply lib/trino-parquet/src/main/java/io/trino/parquet/writer/ParquetDataOutput.java:58]: > That only saves on new array allocations and doesn't avoid copying.

OK. I thought we can avoid copying too by decompressing pages into reusable buffer

Comment (raunaqmorarka) [reply lib/trino-parquet/src/main/java/io/trino/parquet/writer/ParquetDataOutput.java:58]: If the buffer is reused across multiple pages, then I think we have to copy the compressed data out as we accumulate many pages worth of data before flushing to storage.
For avoiding copy, we have to avoid reusing the buffer and be okay with wasted space per page

Comment (sopel39) [reply lib/trino-parquet/src/main/java/io/trino/parquet/writer/ParquetDataOutput.java:58]: > For avoiding copy, we have to avoid reusing the buffer and be okay with wasted space per page

Why such buffer can't be reused? I thought decompressed pages are short lived

Comment (raunaqmorarka) [reply lib/trino-parquet/src/main/java/io/trino/parquet/writer/ParquetDataOutput.java:58]: Compressed pages need to be kept in memory until the row group size threshold is met. So every column writer accumulates many compressed pages before a flush to output stream.
As for the page data before compression, that part is inside parquet-mr classes and not under our control.
In any case, it's hard to make meaningful improvement here while we are tied to using parquet-mr classes, we should be able to do better when we rewrite this fully as trino code.
